worker/codalabworker/local_run/local_run_state.py

```diff
@@ -0,0 +1,277 @@
+    """
+    RUNNING = 'LOCAL_RUN.RUNNING'
+
+    """
+    This stage encompasses cleaning up intermediary components like
+    the dependency symlinks and also the releasing of dependencies
+    """
+    CLEANING = 'LOCAL_RUN.CLEANING'
+
+    """
+    Uploading results means the job's results are getting uploaded to the server
+    """
+    UPLOADING_RESULTS = 'LOCAL_RUN.UPLOADING_RESULTS'
+
+    """
+    Finalizing means the worker is finalizing the bundle metadata with the server
+    """
+    FINALIZING = 'LOCAL_RUN.FINALIZING'
+
+    """
+    Finished means the worker is done with this run
+    """
+    FINISHED = 'LOCAL_RUN.FINISHED'
+
+LocalRunState = namedtuple('RunState',
+    ['stage', 'run_status', 'bundle', 'bundle_path', 'resources', 'start_time',
+'container_id', 'docker_image', 'is_killed', 'has_contents', 'cpuset', 'gpuset', 'info'])
+
+class LocalRunStateMachine(StateTransitioner):
+    """
+    Manages the state machine of the runs running on the local machine
+    """
+
+    def __init__(self, run_manager):
+        super(LocalRunStateMachine, self).__init__()
+        self._run_manager = run_manager
+        self.add_transition(LocalRunStage.PREPARING, self._transition_from_PREPARING)
+        self.add_transition(LocalRunStage.RUNNING, self._transition_from_RUNNING)
+        self.add_transition(LocalRunStage.CLEANING, self._transition_from_CLEANING)
+        self.add_transition(LocalRunStage.UPLOADING_RESULTS, self._transition_from_UPLOADING_RESULTS)
+        self.add_transition(LocalRunStage.FINALIZING, self._transition_from_FINALIZING)
+        self.add_transition(LocalRunStage.FINISHED, self._transition_from_FINISHED)
+
+    def _transition_from_PREPARING(self, run_state):
+        if run_state.is_killed:
+            return run_state._replace(stage=LocalRunStage.CLEANING, container_id=None)
+        # first attempt to get() every dependency/image so that downloads start in parallel
+        for dep in run_state.bundle['dependencies']:
+            dependency = (dep['parent_uuid'], dep['parent_path'])
+            dependency_state = self._run_manager.dependency_manager.get(run_state.bundle['uuid'], dependency)
+        docker_image = run_state.resources['docker_image']
+        image_state = self._run_manager.image_manager.get(docker_image)
+
+        # then inspect the state of every dependency/image to see whether all of them are ready
+        for dep in run_state.bundle['dependencies']:
+            dependency = (dep['parent_uuid'], dep['parent_path'])
+            dependency_state = self._run_manager.dependency_manager.get(run_state.bundle['uuid'], dependency)
+            if dependency_state.stage == DependencyStage.DOWNLOADING:
+                status_message = 'Downloading dependency %s: %s done (archived size)' % (
+                            dep['child_path'], size_str(dependency_state.size_bytes))
+                return run_state._replace(run_status=status_message)
```

Comment:
This is not quite accurate - should concatenate the status messages from all the dependencies and docker image...

Reply:
Ideally yes, but that would make a very long status message. If we had a way to show this status better on the UI instead of displaying a single string, I would prefer to do it this way.

Reply:
Could we append a short summary (e.g., "N others")